Issue model, API 생성

EarlyHail opened the pull request:
### 구현의도
-  Issue model 생성
    SDK에서 보내주는 타입으로 데이터 저장

- Issue API
   Add, Get, Delete API 생성

- Error Handling 생성
   `ctx.throw`를 이용한 error handling

### 기능 흐름도, 클래스 다이어그램(선택사항)
- 

### 사용된 기술(선택사항)
- 

### 리뷰 & 논의사항 & 궁금한점(선택사항)
-

EarlyHail committed:
[feat] issue 모델, add issue api 생성

- issue 모델 생성
  타입은 임시로 name, issue로 생성

- add issue api 생성
  body로 data를 받아서 issue 모델에 데이터 삽입

EarlyHail committed:
[feat] error handling 추가

- error handling 추가
  app.ts에 에러 핸들링 라우팅 로직 추가
  `ctx.throw`를 이용
  ### 추후 에러 핸들링 방식에 대해 상의 필요

- addIssue에 validation error 추가

EarlyHail committed:
[fix] test model 수정

- test model schema 수정
  name 필드 타입 object로 변경

EarlyHail committed:
[fix] add Issue 응답 변경

- add Issue 응답 변경
  기존 : issue list
  변경 : status 200

EarlyHail committed:
[feat] get issue / issues api

- get issue api 생성
  issue 단일 조회 기능
  issue의 `_id` key로 조회하도록 구현

- get issues api 생성
  issue 리스트 조회

EarlyHail committed:
[feat] delete issues api 추가

- 모든 issue 리스트를 삭제하는 api 추가
  `개발용` API 입니다

EarlyHail committed:
[fix] Issue 타입 변경

- Issue 타입 변경
  SDK에서 보내주는 타입 형식으로
  `IssueType`, `Issue Schema`의 타입 변경

- 사용자 ip 관련 확인 필요

EarlyHail committed:
[add] cors 추가

- koa에 cors 추가
  SDK 테스트 port와 api 서버 port가 달라서 cors 추가

EarlyHail committed:
[fix] Issue Type, Schema 오타 수정

- Issue Type, Schema에 `browser` 오타 수정

saeeng reviewed (approved):
아주좋습니다!👍

junsushin-dev commented on src/app.ts:
```diff
@@ -16,9 +15,22 @@ const app = new Koa();
 
 app.use(apiRouter().routes());
 
+/**
+ * @TODO
+ * 에러 코드 분리
+ */
+app.on('error', (err, ctx) => {
+  // 에러 코드
+  console.log(err.status);
+  // 에러 메시지
+  console.log(err.message);
+  // console.log('server error', err, ctx);
+});
+
```
Todo로 주석처리 해놓으신 부분 아주 좋은 것 같습니다! 😊

junsushin-dev commented on src/routes/issue/controllers/getIssue.ts:
```diff
@@ -0,0 +1,10 @@
+import { Context, Next } from 'koa';
+import Issue, { IssueTypeModel } from '../../../models/Issue';
+
+export default async (ctx: Context, next: Next) => {
+  const { issueId } = ctx.params;
+  const result: IssueTypeModel | null = await Issue.findOne({ _id: issueId });
+  ctx.body = result;
+
+  await next();
+};
```
이쪽은 result로 null이 뜰 경우 에러핸들링을 추후에 추가하면 좋을 것 같습니다! 😀

junsushin-dev replied:
app.ts 에서 에러핸들링 부분처럼 //Todo로 적어 놓으면 더 좋을 것 같네요.

EarlyHail replied:
처음에는 간단하게 null을 응답하려고 생각했었는데....
빈 배열을 넘기던지 아니면 다른 방법을 생각하는게 좋겠네요

saeeng replied:
맞는 데이터가 없는 상황이 에러는 아닌것같아서... 저는 빈배열 혹은 null을 반환해주는게 맞을것같아요!

EarlyHail replied:
생각해보니 getIssue라 빈배열은 안되겠네요..... ㅋㅋㅋ 내일 스크럼때 이야기해봐요!

junsushin-dev replied:
저는 단일 issue에 대해 get 접근을 하는 api니까, 찾지 못했을 경우 ctx.throw(400, 'issue with given id not found') 로 처리하고 404 code(Not Found)로 클라이언트에 응답해주면 되지 않을까 생각하고 있었습니다 🤔
getIssue**s**에서는 호진님 말씀처럼 빈 배열로 줘도 괜찮지 않을 까 싶습니다. 😁

junsushin-dev commented on src/routes/issue/controllers/cleanIssues.ts:
```diff
@@ -0,0 +1,8 @@
+import { Context, Next } from 'koa';
+import Issue from '../../../models/Issue';
+
+export default async (ctx: Context, next: Next) => {
+  await Issue.deleteMany({});
+  ctx.response.status = 200;
+  await next();
+};
```
Issue 데이터를 전부 밀어버리는 API 같은데, 테스트 용도로 있는 것일까요? 
실제 배포 서버에서는 위험하니 막아놓아야 할 것 같고, 현재 테스트 용도라면 주석으로 '// Temp - 개발용' 이런식으로 명시해주시면 더 좋을 것 같습니다! 😉

junsushin-dev commented on src/routes/issue/controllers/addIssue.ts:
```diff
@@ -0,0 +1,16 @@
+
+export default async (ctx: Context, next: Next) => {
+  const newIssue: IssueType = ctx.request.body;
+  newIssue.meta.ip = ctx.header.host;
```
ctx.header.host를 ctx.ip로 바꾸어야 하지 않을까요? 🤔

junsushin-dev replied:
로컬에서 테스트해보니까 현재 host url이 들어오고 있는 것 같습니다. 
POSTMAN에서 127.0.0.1:3000로 보내면 127.0.0.1:3000이 들어오고, localhost:3000으로 보내면 localhost:3000이 들어오고 있는 것 같네요.

EarlyHail replied:
현재 사용자의 ip를 얻어오는걸 정확히 테스트할수 없어서 이슈에 등록해놨습니다.

추후 개발 서버에 배포해보고 확인하려고 이슈에 등록해놨습니다.

junsushin-dev replied:
아하 아까 로컬 환경이라서 ip 관련해서 이야기하셨던 것이 이 부분이었군요! 피드백과 빠른 이슈 등록 감사합니다!!
이런 부분 바로 이슈로 등록해주는 것은 정말 좋은 것 같습니다. 😎

junsushin-dev commented on src/models/Issue.ts:
```diff
@@ -0,0 +1,62 @@
+  message: { type: String, required: true },
+  stack: [
+    {
+      columnNo: { type: String, required: true },
+      lineNo: { type: String, required: true },
```
line, column no는 나중에 혹시 SDK에서 보내온 소스코드나 GitHub에서 코드를 읽어올 때 사용할 수 있을 것 같아서 String type이 아니라 Number type으로 변환해서 저장하는 것이 좋지 않을까요? 다른 분들은 어떻게 생각하시는지 궁금합니다. 📌

junsushin-dev reviewed (commented):
안녕하세요, 리뷰 코멘트 몇 개 남겼습니다! 검토 부탁드립니다 :)

EarlyHail committed:
[refactor] TODO 주석 추가

- `getIssue` controller에 null 처리 todo 추가

- `cleanIssue` api에 주석 추가
  추후에 삭제하기 쉽도록
  `개발용`, `warning` 주석 추가

juyoungpark718 commented on src/models/Issue.ts:
```diff
@@ -0,0 +1,21 @@
+
+const Issue = model<IssueTypeModel>('Issue', issueSchema);
+
+export const build = (issue: IssueType): IssueTypeModel => {
+  return new Issue(issue);
+};
```
Issue.build(attr:IssueType)
이런 식으로 사용할 수 있는 방법이 있었는데, 후에 리팩토링 시도해보면 좋을 것 같습니다 👍

junsushin-dev replied:
저도 주영님 말씀처럼 build가 혹시 나중에 여러 모델에 대해서 쓰일 수 있으니, Issue.build나 buildIssue와 같은 방식으로 추후 리팩토링하면 더 좋을 것 같네요! 💫

juyoungpark718 reviewed (approved):
너무 고생하셨습니다 !! 👍

junsushin-dev reviewed (approved):
호진님 세웅님 정말 고생하셨습니다!! 혹~시 제가 너무 사소한 부분에 대해 까다롭게 보는 것 같으면 언제든지 말씀주세요! 😌
